File: day_22_OSI_models/day_22_OSI_models.cpp
/*
 * OSI Model: Industry-Style C++17 Case Study
 *
 * Scenario:
 * A small enterprise network contains client devices, an access switch,
 * a router, an application server, and a monitoring component.
 *
 * The program models packet construction, MAC learning, routing,
 * transport connections, application requests, diagnostics, validation,
 * security controls, and performance metrics.
 *
 * Compile:
 *   g++ -std=c++17 -O2 osi_models.cpp -o osi_models
 *
 * Run:
 *   ./osi_models
 */

#include <algorithm>
#include <array>
#include <chrono>
#include <cstdint>
#include <iomanip>
#include <iostream>
#include <map>
#include <optional>
#include <sstream>
#include <stdexcept>
#include <string>
#include <unordered_map>
#include <utility>
#include <vector>

using namespace std;

enum class OSILayer {
    Physical = 1,
    DataLink = 2,
    Network = 3,
    Transport = 4,
    Session = 5,
    Presentation = 6,
    Application = 7
};

string layerName(OSILayer layer) {
    switch (layer) {
        case OSILayer::Physical: return "Physical";
        case OSILayer::DataLink: return "Data Link";
        case OSILayer::Network: return "Network";
        case OSILayer::Transport: return "Transport";
        case OSILayer::Session: return "Session";
        case OSILayer::Presentation: return "Presentation";
        case OSILayer::Application: return "Application";
    }

    return "Unknown";
}

string pduName(OSILayer layer) {
    switch (layer) {
        case OSILayer::Physical: return "Bits";
        case OSILayer::DataLink: return "Frame";
        case OSILayer::Network: return "Packet";
        case OSILayer::Transport: return "Segment/Datagram";
        case OSILayer::Session:
        case OSILayer::Presentation:
        case OSILayer::Application:
            return "Data";
    }

    return "Unknown";
}

void title(const string& text) {
    cout << "\n" << string(78, '=') << "\n";
    cout << text << "\n";
    cout << string(78, '=') << "\n";
}

class IPv4Address {
private:
    uint32_t value;

public:
    explicit IPv4Address(uint32_t value = 0) : value(value) {}

    static IPv4Address parse(const string& text) {
        unsigned int a, b, c, d;
        char first, second, third;

        stringstream stream(text);

        if (!(stream >> a >> first >> b >> second >> c >> third >> d) ||
            first != '.' || second != '.' || third != '.' ||
            a > 255 || b > 255 || c > 255 || d > 255) {
            throw invalid_argument("Invalid IPv4 address: " + text);
        }

        uint32_t value =
            (a << 24) |
            (b << 16) |
            (c << 8) |
            d;

        return IPv4Address(value);
    }

    uint32_t raw() const {
        return value;
    }

    string toString() const {
        ostringstream output;

        output
            << ((value >> 24) & 255) << "."
            << ((value >> 16) & 255) << "."
            << ((value >> 8) & 255) << "."
            << (value & 255);

        return output.str();
    }

    bool operator==(const IPv4Address& other) const {
        return value == other.value;
    }
};

class CIDRNetwork {
private:
    IPv4Address networkAddress;
    int prefixLength;

public:
    CIDRNetwork(IPv4Address address, int prefix)
        : networkAddress(address), prefixLength(prefix) {

        if (prefix < 0 || prefix > 32) {
            throw invalid_argument("CIDR prefix must be between 0 and 32.");
        }
    }

    bool contains(IPv4Address address) const {
        if (prefixLength == 0) {
            return true;
        }

        uint32_t mask = 0xFFFFFFFFu << (32 - prefixLength);

        return (address.raw() & mask) ==
               (networkAddress.raw() & mask);
    }

    int prefix() const {
        return prefixLength;
    }

    string toString() const {
        return networkAddress.toString() + "/" +
               to_string(prefixLength);
    }
};

struct EthernetFrame {
    string sourceMAC;
    string destinationMAC;
    uint16_t etherType;
    string payload;
};

struct IPPacket {
    IPv4Address source;
    IPv4Address destination;
    uint8_t ttl;
    uint8_t protocol;
    string payload;

    bool forward() {
        if (ttl == 0) {
            return false;
        }

        --ttl;
        return ttl > 0;
    }
};

struct TCPSegment {
    uint16_t sourcePort;
    uint16_t destinationPort;
    uint32_t sequenceNumber;
    uint32_t acknowledgmentNumber;
    vector<string> flags;
    string payload;
};

struct Route {
    CIDRNetwork network;
    string nextHop;

    Route(CIDRNetwork network, string nextHop)
        : network(std::move(network)),
          nextHop(std::move(nextHop)) {}
};

class Router {
private:
    vector<Route> routingTable;

public:
    void addRoute(const CIDRNetwork& network, const string& nextHop) {
        routingTable.emplace_back(network, nextHop);
    }

    const Route* lookup(const IPv4Address& destination) const {
        const Route* bestRoute = nullptr;

        for (const auto& route : routingTable) {
            if (!route.network.contains(destination)) {
                continue;
            }

            if (bestRoute == nullptr ||
                route.network.prefix() > bestRoute->network.prefix()) {
                bestRoute = &route;
            }
        }

        return bestRoute;
    }

    void showTable() const {
        for (const auto& route : routingTable) {
            cout << left
                 << setw(20) << route.network.toString()
                 << " -> " << route.nextHop << "\n";
        }
    }
};

class EthernetSwitch {
private:
    unordered_map<string, string> macTable;

public:
    void learn(const string& mac, const string& port) {
        macTable[mac] = port;
    }

    string forward(const string& destinationMAC) const {
        auto iterator = macTable.find(destinationMAC);

        if (iterator == macTable.end()) {
            return "FLOOD";
        }

        return iterator->second;
    }

    void showMACTable() const {
        for (const auto& [mac, port] : macTable) {
            cout << left << setw(25) << mac
                 << " -> " << port << "\n";
        }
    }
};

class TransportConnection {
private:
    uint16_t sourcePort;
    uint16_t destinationPort;
    uint32_t nextSequenceNumber;
    bool established;

public:
    TransportConnection(
        uint16_t sourcePort,
        uint16_t destinationPort
    )
        : sourcePort(sourcePort),
          destinationPort(destinationPort),
          nextSequenceNumber(10000),
          established(false) {}

    void establish() {
        cout << "TCP SYN: "
             << sourcePort << " -> "
             << destinationPort << "\n";

        cout << "TCP SYN-ACK: "
             << destinationPort << " -> "
             << sourcePort << "\n";

        cout << "TCP ACK: "
             << sourcePort << " -> "
             << destinationPort << "\n";

        established = true;
    }

    void send(const string& data) {
        if (!established) {
            throw runtime_error(
                "Cannot send application data before TCP establishment."
            );
        }

        cout << "TCP segment seq=" << nextSequenceNumber
             << " payload=" << data << "\n";

        nextSequenceNumber += static_cast<uint32_t>(data.size());
    }
};

class SecurityPolicy {
private:
    map<OSILayer, string> controls;

public:
    SecurityPolicy() {
        controls[OSILayer::Physical] =
            "Physical access control";
        controls[OSILayer::DataLink] =
            "VLAN and 802.1X controls";
        controls[OSILayer::Network] =
            "ACLs and segmentation";
        controls[OSILayer::Transport] =
            "Port and stateful firewall controls";
        controls[OSILayer::Session] =
            "Session timeout and state validation";
        controls[OSILayer::Presentation] =
            "TLS and certificate validation";
        controls[OSILayer::Application] =
            "Authentication, authorization, input validation";
    }

    void display() const {
        for (const auto& [layer, control] : controls) {
            cout << "L" << static_cast<int>(layer)
                 << " " << setw(14) << left << layerName(layer)
                 << " : " << control << "\n";
        }
    }
};

struct DiagnosticEvent {
    OSILayer layer;
    string action;
    string details;
};

class NetworkMonitor {
private:
    vector<DiagnosticEvent> events;

public:
    void record(
        OSILayer layer,
        const string& action,
        const string& details
    ) {
        events.push_back({layer, action, details});
    }

    void display() const {
        for (const auto& event : events) {
            cout << "L"
                 << static_cast<int>(event.layer)
                 << " "
                 << setw(15)
                 << left
                 << layerName(event.layer)
                 << setw(14)
                 << event.action
                 << event.details
                 << "\n";
        }
    }
};

class EnterpriseNetwork {
private:
    EthernetSwitch accessSwitch;
    Router router;
    NetworkMonitor monitor;
    SecurityPolicy securityPolicy;

    const string clientMAC = "AA:BB:CC:DD:EE:01";
    const string gatewayMAC = "AA:BB:CC:DD:EE:FE";

    IPv4Address clientIP = IPv4Address::parse("192.168.10.20");
    IPv4Address gatewayIP = IPv4Address::parse("192.168.10.1");
    IPv4Address serverIP = IPv4Address::parse("10.20.30.50");

public:
    EnterpriseNetwork() {
        /*
         * Layer 2 configuration.
         * The switch learns which MAC addresses appear on which ports.
         */
        accessSwitch.learn(clientMAC, "Access-Port-1");
        accessSwitch.learn(gatewayMAC, "Uplink-Port-24");

        /*
         * Layer 3 routing configuration.
         * Longest-prefix matching will select the most specific route.
         */
        router.addRoute(
            CIDRNetwork(
                IPv4Address::parse("192.168.10.0"),
                24
            ),
            "Local LAN"
        );

        router.addRoute(
            CIDRNetwork(
                IPv4Address::parse("10.20.30.0"),
                24
            ),
            "Application Network"
        );

        router.addRoute(
            CIDRNetwork(
                IPv4Address::parse("0.0.0.0"),
                0
            ),
            "Internet Gateway"
        );
    }

    void printArchitecture() const {
        title("Enterprise Network Architecture");

        cout << "Client\n";
        cout << "  |\n";
        cout << "Access Switch\n";
        cout << "  |\n";
        cout << "Router\n";
        cout << "  |\n";
        cout << "Application Server\n";

        cout << "\nClient IP : " << clientIP.toString() << "\n";
        cout << "Server IP : " << serverIP.toString() << "\n";
    }

    void demonstrateLayer2() {
        title("Layer 2: Switching");

        cout << "MAC table:\n";
        accessSwitch.showMACTable();

        cout << "\nDestination MAC "
             << gatewayMAC
             << " -> "
             << accessSwitch.forward(gatewayMAC)
             << "\n";
    }

    void demonstrateLayer3() {
        title("Layer 3: Routing");

        cout << "Routing table:\n";
        router.showTable();

        const Route* route = router.lookup(serverIP);

        if (route == nullptr) {
            throw runtime_error("No route to application server.");
        }

        cout << "\nDestination "
             << serverIP.toString()
             << " uses "
             << route->network.toString()
             << " -> "
             << route->nextHop
             << "\n";
    }

    void demonstrateLayer4() {
        title("Layer 4: Transport");

        TransportConnection connection(51500, 443);
        connection.establish();
        connection.send("GET /inventory HTTP/1.1");
    }

    void demonstrateEndToEndRequest() {
        title("End-to-End HTTPS Request");

        monitor.record(
            OSILayer::Application,
            "CREATE",
            "Browser creates HTTP request"
        );

        monitor.record(
            OSILayer::Presentation,
            "PROTECT",
            "TLS encrypts application communication"
        );

        monitor.record(
            OSILayer::Session,
            "TRACK",
            "Logical session state is maintained"
        );

        monitor.record(
            OSILayer::Transport,
            "SEGMENT",
            "TCP targets port 443"
        );

        monitor.record(
            OSILayer::Network,
            "ROUTE",
            "IP packet targets application server"
        );

        monitor.record(
            OSILayer::DataLink,
            "FRAME",
            "Ethernet frame targets next hop"
        );

        monitor.record(
            OSILayer::Physical,
            "TRANSMIT",
            "Bits become electrical/optical/radio signals"
        );

        monitor.display();
    }

    void demonstrateTTL() {
        title("Failure Case: TTL Expiration");

        IPPacket packet{
            clientIP,
            serverIP,
            2,
            6,
            "HTTPS"
        };

        cout << "Initial TTL: " << static_cast<int>(packet.ttl) << "\n";

        bool forwarded = packet.forward();
        cout << "After router 1: TTL="
             << static_cast<int>(packet.ttl)
             << ", forwarded="
             << boolalpha
             << forwarded
             << "\n";

        forwarded = packet.forward();
        cout << "After router 2: TTL="
             << static_cast<int>(packet.ttl)
             << ", forwarded="
             << boolalpha
             << forwarded
             << "\n";

        if (!forwarded) {
            cout << "Packet expired; a real router would normally generate "
                    "an ICMP Time Exceeded response.\n";
        }
    }

    void demonstrateSecurity() const {
        title("Security Controls Across Layers");
        securityPolicy.display();
    }

    void demonstrateTroubleshooting() const {
        title("Layer-Oriented Troubleshooting");

        const vector<pair<OSILayer, string>> symptoms = {
            {OSILayer::Physical, "No link or damaged cable"},
            {OSILayer::DataLink, "VLAN or MAC-learning problem"},
            {OSILayer::Network, "Incorrect address or missing route"},
            {OSILayer::Transport, "Port blocked or handshake failure"},
            {OSILayer::Session, "Expired or invalid session state"},
            {OSILayer::Presentation, "TLS/certificate/encoding failure"},
            {OSILayer::Application, "HTTP or application-level error"}
        };

        for (const auto& [layer, symptom] : symptoms) {
            cout << "L"
                 << static_cast<int>(layer)
                 << " "
                 << setw(15)
                 << left
                 << layerName(layer)
                 << ": "
                 << symptom
                 << "\n";
        }
    }
};

void demonstrateOSIReference() {
    title("OSI Seven-Layer Reference");

    for (int i = 1; i <= 7; ++i) {
        OSILayer layer = static_cast<OSILayer>(i);

        cout << "L"
             << i
             << " "
             << setw(15)
             << left
             << layerName(layer)
             << " PDU="
             << pduName(layer)
             << "\n";
    }
}

void demonstrateComplexity() {
    title("Algorithmic Complexity");

    cout << "MAC-table lookup with unordered_map: expected O(1)\n";
    cout << "Route lookup in this demonstration: O(R), where R is route count.\n";
    cout << "Longest-prefix selection scans all candidate routes.\n";
    cout << "Production routers use optimized forwarding structures to reduce lookup cost.\n";
}

void demonstrateTradeoffs() {
    title("Architectural Trade-offs");

    cout << "Layering benefits:\n";
    cout << "  - Separation of concerns\n";
    cout << "  - Interoperability\n";
    cout << "  - Easier diagnostics\n";
    cout << "  - Independent protocol evolution\n";

    cout << "\nLayering trade-offs:\n";
    cout << "  - Additional headers and processing\n";
    cout << "  - Real protocols may span conceptual layers\n";
    cout << "  - Multiple layers may independently implement similar controls\n";
}

void runValidationTests() {
    title("Validation Tests");

    IPv4Address address = IPv4Address::parse("192.168.1.10");

    if (address.toString() != "192.168.1.10") {
        throw runtime_error("IPv4 conversion test failed.");
    }

    CIDRNetwork network(
        IPv4Address::parse("192.168.1.0"),
        24
    );

    if (!network.contains(address)) {
        throw runtime_error("CIDR containment test failed.");
    }

    if (network.contains(IPv4Address::parse("10.0.0.1"))) {
        throw runtime_error("CIDR exclusion test failed.");
    }

    bool exceptionCaught = false;

    try {
        IPv4Address::parse("999.1.1.1");
    } catch (const invalid_argument&) {
        exceptionCaught = true;
    }

    if (!exceptionCaught) {
        throw runtime_error("Invalid IP test failed.");
    }

    cout << "All validation tests passed.\n";
}

int main() {
    try {
        title("OSI MODEL C++ INDUSTRY-STYLE CASE STUDY");

        demonstrateOSIReference();

        EnterpriseNetwork network;

        network.printArchitecture();
        network.demonstrateLayer2();
        network.demonstrateLayer3();
        network.demonstrateLayer4();
        network.demonstrateEndToEndRequest();
        network.demonstrateTTL();
        network.demonstrateSecurity();
        network.demonstrateTroubleshooting();

        demonstrateComplexity();
        demonstrateTradeoffs();
        runValidationTests();

        title("CASE STUDY COMPLETE");
        cout << "Enterprise network simulation completed successfully.\n";

        return 0;
    } catch (const exception& error) {
        cerr << "Fatal error: " << error.what() << "\n";
        return 1;
    }
}

File: day_22_OSI_models/day_22_OSI_models.js
/*
 * OSI Model: JavaScript Study Implementation
 *
 * This file complements the Python implementation by using JavaScript
 * objects, classes, Maps, asynchronous functions, validation, event-driven
 * behavior, and packet-processing simulations.
 *
 * Run with:
 *   node osi_models.js
 */

"use strict";

const OSILayer = Object.freeze({
    PHYSICAL: 1,
    DATA_LINK: 2,
    NETWORK: 3,
    TRANSPORT: 4,
    SESSION: 5,
    PRESENTATION: 6,
    APPLICATION: 7
});

const LAYER_NAMES = Object.freeze({
    1: "Physical",
    2: "Data Link",
    3: "Network",
    4: "Transport",
    5: "Session",
    6: "Presentation",
    7: "Application"
});

const PDU_NAMES = Object.freeze({
    1: "Bits",
    2: "Frame",
    3: "Packet",
    4: "Segment / Datagram",
    5: "Data",
    6: "Data",
    7: "Data"
});

function printTitle(title) {
    console.log(`\n${"=".repeat(78)}`);
    console.log(title);
    console.log("=".repeat(78));
}

function demonstrateLayers() {
    printTitle("1. Seven OSI Layers");

    const layers = [
        [1, "Physical", "Bits", "Signals, cables, fiber, radio"],
        [2, "Data Link", "Frame", "Ethernet, Wi-Fi, MAC addressing"],
        [3, "Network", "Packet", "IPv4, IPv6, routing"],
        [4, "Transport", "Segment/Datagram", "TCP, UDP, ports"],
        [5, "Session", "Data", "Session/dialog management"],
        [6, "Presentation", "Data", "Encoding, compression, encryption"],
        [7, "Application", "Data", "HTTP, DNS, SMTP, SSH"]
    ];

    for (const [number, name, pdu, examples] of layers) {
        console.log(
            `L${number} ${name.padEnd(15)} PDU=${pdu.padEnd(17)} ${examples}`
        );
    }
}

function demonstrateEncapsulation() {
    printTitle("2. Encapsulation");

    const applicationData = "GET /index.html HTTP/1.1";
    const tlsData = `TLS(${applicationData})`;
    const tcpSegment = `TCP[51500 -> 443] ${tlsData}`;
    const ipPacket = `IP[192.168.1.20 -> 203.0.113.10] ${tcpSegment}`;
    const ethernetFrame = `ETH[AA:BB:CC:DD:EE:01 -> AA:BB:CC:DD:EE:FE] ${ipPacket}`;

    console.log("Application:", applicationData);
    console.log("Presentation:", tlsData);
    console.log("Transport:", tcpSegment);
    console.log("Network:", ipPacket);
    console.log("Data Link:", ethernetFrame);
    console.log("Physical: binary signals representing the frame");

    console.log("\nDecapsulation removes these wrappers in reverse order.");
}

class TCPSegment {
    constructor(sourcePort, destinationPort, sequenceNumber, acknowledgmentNumber, flags, payload = "") {
        this.sourcePort = sourcePort;
        this.destinationPort = destinationPort;
        this.sequenceNumber = sequenceNumber;
        this.acknowledgmentNumber = acknowledgmentNumber;
        this.flags = flags;
        this.payload = payload;
    }

    describe() {
        return {
            ports: `${this.sourcePort} -> ${this.destinationPort}`,
            sequenceNumber: this.sequenceNumber,
            acknowledgmentNumber: this.acknowledgmentNumber,
            flags: this.flags.join(", "),
            payload: this.payload
        };
    }
}

function demonstrateTCPHandshake() {
    printTitle("3. TCP Three-Way Handshake");

    const clientPort = 51500;
    const serverPort = 443;
    const clientISN = 10000;
    const serverISN = 70000;

    const handshake = [
        new TCPSegment(clientPort, serverPort, clientISN, 0, ["SYN"]),
        new TCPSegment(serverPort, clientPort, serverISN, clientISN + 1, ["SYN", "ACK"]),
        new TCPSegment(clientPort, serverPort, clientISN + 1, serverISN + 1, ["ACK"])
    ];

    handshake.forEach((segment, index) => {
        console.log(`Message ${index + 1}:`, segment.describe());
    });
}

function demonstrateAddressing() {
    printTitle("4. Addressing");

    const macAddress = "AA:BB:CC:DD:EE:01";
    const ipAddress = "192.168.1.20";
    const destinationPort = 443;

    console.log("Layer 2 MAC:", macAddress);
    console.log("Layer 3 IP:", ipAddress);
    console.log("Layer 4 port:", destinationPort);

    console.log("\nA complete endpoint can be represented as:");
    console.log(`${ipAddress}:${destinationPort}`);
}

function ipv4ToInteger(address) {
    const octets = address.split(".").map(Number);

    if (
        octets.length !== 4 ||
        octets.some(
            octet => !Number.isInteger(octet) || octet < 0 || octet > 255
        )
    ) {
        throw new Error(`Invalid IPv4 address: ${address}`);
    }

    return (
        (((octets[0] << 24) >>> 0) |
            (octets[1] << 16) |
            (octets[2] << 8) |
            octets[3]) >>> 0
    );
}

function isValidPort(port) {
    return Number.isInteger(port) && port >= 1 && port <= 65535;
}

function demonstrateValidation() {
    printTitle("5. Network Validation");

    const addresses = [
        "192.168.1.1",
        "10.0.0.5",
        "256.1.1.1",
        "192.168.1"
    ];

    for (const address of addresses) {
        try {
            const value = ipv4ToInteger(address);
            console.log(`${address.padEnd(16)} valid -> ${value}`);
        } catch (error) {
            console.log(`${address.padEnd(16)} invalid -> ${error.message}`);
        }
    }

    const ports = [0, 22, 443, 65535, 65536];

    for (const port of ports) {
        console.log(`Port ${port}: ${isValidPort(port) ? "valid" : "invalid"}`);
    }
}

class EthernetSwitch {
    constructor() {
        this.macTable = new Map();
    }

    learn(macAddress, port) {
        this.macTable.set(macAddress.toUpperCase(), port);
    }

    forward(destinationMac) {
        const normalized = destinationMac.toUpperCase();
        return this.macTable.get(normalized) ?? "FLOOD";
    }
}

function demonstrateSwitching() {
    printTitle("6. Layer 2 Switching");

    const switchDevice = new EthernetSwitch();

    switchDevice.learn("AA:AA:AA:AA:AA:10", "Port 1");
    switchDevice.learn("AA:AA:AA:AA:AA:20", "Port 2");
    switchDevice.learn("AA:AA:AA:AA:AA:30", "Port 3");

    console.log(
        "Known destination:",
        switchDevice.forward("AA:AA:AA:AA:AA:20")
    );

    console.log(
        "Unknown destination:",
        switchDevice.forward("AA:AA:AA:AA:AA:99")
    );
}

class RouteTable {
    constructor() {
        this.routes = [];
    }

    addRoute(networkAddress, prefixLength, nextHop) {
        if (
            !Number.isInteger(prefixLength) ||
            prefixLength < 0 ||
            prefixLength > 32
        ) {
            throw new Error("IPv4 prefix length must be between 0 and 32.");
        }

        this.routes.push({
            networkAddress,
            prefixLength,
            nextHop
        });
    }

    /*
     * This simplified matcher converts IPv4 addresses to integers and applies
     * the CIDR mask. Longest-prefix matching selects the most specific route.
     */
    lookup(destination) {
        const destinationInteger = ipv4ToInteger(destination);

        const matches = this.routes.filter(route => {
            const networkInteger = ipv4ToInteger(route.networkAddress);

            if (route.prefixLength === 0) {
                return true;
            }

            const mask =
                (0xffffffff << (32 - route.prefixLength)) >>> 0;

            return (
                (destinationInteger & mask) >>> 0 ===
                (networkInteger & mask) >>> 0
            );
        });

        if (matches.length === 0) {
            return null;
        }

        return matches.reduce((best, current) =>
            current.prefixLength > best.prefixLength ? current : best
        );
    }
}

function demonstrateRouting() {
    printTitle("7. Layer 3 Routing");

    const routingTable = new RouteTable();

    routingTable.addRoute("0.0.0.0", 0, "Internet Gateway");
    routingTable.addRoute("10.0.0.0", 8, "Private Router");
    routingTable.addRoute("10.20.0.0", 16, "Data Center Router");
    routingTable.addRoute("10.20.30.0", 24, "Application Router");

    const destination = "10.20.30.44";
    const route = routingTable.lookup(destination);

    console.log(`Destination: ${destination}`);
    console.log("Selected route:", route);
}

function demonstrateTransportComparison() {
    printTitle("8. TCP and UDP");

    const comparison = [
        ["Connection", "TCP-oriented connection", "UDP connectionless"],
        ["Reliability", "Acknowledgments/retransmission", "Not built into UDP"],
        ["Ordering", "Ordered byte stream", "No ordering guarantee"],
        ["Overhead", "Higher", "Lower"],
        ["Typical use", "HTTPS, SSH", "DNS, real-time protocols"]
    ];

    console.table(
        comparison.map(([property, tcp, udp]) => ({
            Property: property,
            TCP: tcp,
            UDP: udp
        }))
    );
}

function demonstrateDNS() {
    printTitle("9. DNS");

    const dnsCache = new Map([
        ["example.com", "93.184.216.34"],
        ["localhost", "127.0.0.1"]
    ]);

    const domain = "example.com";
    const address = dnsCache.get(domain);

    console.log(`Query: ${domain}`);

    if (address) {
        console.log(`Answer: ${address}`);
    } else {
        console.log("No cached answer; a DNS resolver would perform a lookup.");
    }
}

function demonstrateHTTP() {
    printTitle("10. HTTP");

    const request = [
        "GET /index.html HTTP/1.1",
        "Host: example.com",
        "Accept: text/html",
        "Connection: close",
        ""
    ].join("\r\n");

    console.log(request);

    console.log("\nTypical HTTPS stack:");
    console.log("HTTP -> TLS -> TCP -> IP -> Ethernet/Wi-Fi -> Physical medium");
}

function demonstrateModernStack() {
    printTitle("11. Modern Protocol Stacks");

    console.log("Traditional HTTPS:");
    console.log("HTTP -> TLS -> TCP -> IP -> Link");

    console.log("\nHTTP/3:");
    console.log("HTTP/3 -> QUIC -> UDP -> IP -> Link");

    console.log(
        "\nThe OSI model remains useful as a conceptual framework even when"
    );
    console.log(
        "modern protocols combine responsibilities that do not fit one layer exactly."
    );
}

class PacketTracer {
    constructor() {
        this.events = [];
    }

    record(layer, action, details) {
        this.events.push({
            layer,
            layerName: LAYER_NAMES[layer],
            action,
            details,
            timestamp: Date.now()
        });
    }

    print() {
        for (const event of this.events) {
            console.log(
                `L${event.layer} ${event.layerName.padEnd(14)} ` +
                `${event.action.padEnd(12)} ${event.details}`
            );
        }
    }
}

function demonstratePacketTrace() {
    printTitle("12. End-to-End Packet Trace");

    const tracer = new PacketTracer();

    tracer.record(7, "CREATE", "Browser creates HTTPS request");
    tracer.record(6, "PROTECT", "TLS protects application communication");
    tracer.record(5, "TRACK", "Session state is maintained");
    tracer.record(4, "SEGMENT", "TCP targets destination port 443");
    tracer.record(3, "ROUTE", "IP packet is routed toward remote network");
    tracer.record(2, "FRAME", "Frame targets the local next hop");
    tracer.record(1, "TRANSMIT", "Bits/signals travel over the medium");

    tracer.print();
}

function demonstrateAsyncNetworkBehavior() {
    printTitle("13. Asynchronous Network Behavior");

    /*
     * JavaScript's event-driven model is useful for demonstrating how an
     * application can start I/O and continue processing other work while
     * awaiting a result.
     */
    const simulatedNetworkRequest = new Promise(resolve => {
        setTimeout(() => {
            resolve({
                status: 200,
                layer: OSILayer.APPLICATION,
                message: "HTTP response received"
            });
        }, 100);
    });

    return simulatedNetworkRequest.then(response => {
        console.log("Asynchronous result:", response);
    });
}

function demonstrateSubnetConcept() {
    printTitle("14. CIDR Subnetting");

    const examples = [
        { network: "192.168.1.0", prefix: 24 },
        { network: "192.168.1.0", prefix: 26 },
        { network: "10.0.0.0", prefix: 8 }
    ];

    for (const item of examples) {
        const totalAddresses = 2 ** (32 - item.prefix);

        console.log(
            `${item.network}/${item.prefix}: ` +
            `${totalAddresses} total IPv4 addresses`
        );

        if (totalAddresses >= 4) {
            console.log(
                `Traditional usable-host count: ${totalAddresses - 2}`
            );
        }
    }
}

function demonstrateSecurity() {
    printTitle("15. Security Across OSI Layers");

    const controls = {
        1: "Physical access control and medium protection",
        2: "VLAN security, 802.1X, switch protections",
        3: "ACLs, segmentation, IPsec, routing security",
        4: "Firewalls and transport-port controls",
        5: "Session expiration and state validation",
        6: "TLS and certificate validation",
        7: "Authentication, authorization, input validation"
    };

    for (let layer = 1; layer <= 7; layer++) {
        console.log(
            `L${layer} ${LAYER_NAMES[layer].padEnd(14)} ${controls[layer]}`
        );
    }
}

function demonstratePerformance() {
    printTitle("16. Performance Considerations");

    const metrics = {
        "Physical": "Bandwidth, interference, signal quality",
        "Data Link": "Frame overhead and switching",
        "Network": "Routing, packet loss, MTU",
        "Transport": "RTT, congestion, retransmission",
        "Session": "Connection/session setup",
        "Presentation": "Encryption and compression cost",
        "Application": "Serialization and server processing"
    };

    for (const [layer, metric] of Object.entries(metrics)) {
        console.log(`${layer.padEnd(15)} ${metric}`);
    }
}

function demonstrateCommonMistakes() {
    printTitle("17. Common OSI Misconceptions");

    const mistakes = [
        "OSI is not a literal implementation required by every network stack.",
        "A protocol may perform responsibilities associated with multiple layers.",
        "A successful ping does not prove that an HTTP service is healthy.",
        "A MAC address and an IP address serve different addressing purposes.",
        "TCP delivery does not guarantee that an application successfully processed data.",
        "Security controls often cross OSI layer boundaries."
    ];

    mistakes.forEach((mistake, index) => {
        console.log(`${index + 1}. ${mistake}`);
    });
}

function demonstrateEventDrivenLayerProcessing() {
    printTitle("18. Event-Driven Packet Processing");

    const listeners = new Map();

    function on(eventName, handler) {
        if (!listeners.has(eventName)) {
            listeners.set(eventName, []);
        }
        listeners.get(eventName).push(handler);
    }

    function emit(eventName, data) {
        for (const handler of listeners.get(eventName) ?? []) {
            handler(data);
        }
    }

    on("frameReceived", frame => {
        console.log("Layer 2 received:", frame);
        emit("packetReady", {
            sourceIP: "192.168.1.20",
            destinationIP: "203.0.113.10"
        });
    });

    on("packetReady", packet => {
        console.log("Layer 3 forwarding decision:", packet);
        emit("segmentReady", {
            sourcePort: 51500,
            destinationPort: 443
        });
    });

    on("segmentReady", segment => {
        console.log("Layer 4 delivered to port:", segment.destinationPort);
    });

    emit("frameReceived", {
        sourceMAC: "AA:BB:CC:DD:EE:01",
        destinationMAC: "AA:BB:CC:DD:EE:FE"
    });
}

function runAssertions() {
    printTitle("19. Self-Tests");

    console.assert(ipv4ToInteger("127.0.0.1") === 2130706433);
    console.assert(isValidPort(443));
    console.assert(!isValidPort(0));
    console.assert(!isValidPort(65536));

    const table = new RouteTable();
    table.addRoute("0.0.0.0", 0, "Default");
    table.addRoute("10.0.0.0", 8, "Private");
    table.addRoute("10.20.30.0", 24, "Application");

    const result = table.lookup("10.20.30.44");
    console.assert(result.nextHop === "Application");

    console.log("All JavaScript self-tests passed.");
}

async function main() {
    printTitle("OSI MODEL JAVASCRIPT STUDY PROGRAM");

    demonstrateLayers();
    demonstrateEncapsulation();
    demonstrateTCPHandshake();
    demonstrateAddressing();
    demonstrateValidation();
    demonstrateSwitching();
    demonstrateRouting();
    demonstrateTransportComparison();
    demonstrateDNS();
    demonstrateHTTP();
    demonstrateModernStack();
    demonstratePacketTrace();
    await demonstrateAsyncNetworkBehavior();
    demonstrateSubnetConcept();
    demonstrateSecurity();
    demonstratePerformance();
    demonstrateCommonMistakes();
    demonstrateEventDrivenLayerProcessing();
    runAssertions();

    printTitle("END OF OSI MODEL JAVASCRIPT STUDY PROGRAM");
}

main().catch(error => {
    console.error("Program failed:", error.message);
    process.exitCode = 1;
});
